### PR TITLE
Add artifacts for email notifications and making the configurations tenantwise

### PR DESCRIPTION
**Pull request by nadundesilva**
## Purpose

In the previous implementation, the users were forced to reset the password in the authentication flow after a globally configurable number of days. This has the following drawbacks.
- Users who had not logged in for a long time won't be prompted to update the credentials.
- Tenant-specific configurations (Number of days after which the passwords should expire) were not possible.

## Goals

- Users will be sent email reminders to update the credentials.
- The configurations were moved to the Resident IdP Properties which makes it a tenant-wise configuration.

## Approach

Using the data collected by the audit data publishers in IS Analytics, a siddhi trigger is used to identify users who need to be reminded and email notifications are sent to them. Without the ID Analytics instance, email reminders feature cannot be used. The password policy authenticator can be used without sending email notifications as well.

The number of days after which the password should expire had been moved to the Resident IdP configurations to make it a tenant-wise configuration.
password-expiry-policy-config

## User stories

- As an admin, I need to be able to automatically notify the users to reset their passwords according to the password policy.
- As an admin, I need to be able to configure the number of days after which the passwords should expire for each tenant separately.

## Release note

N/A

## Documentation

Included with the PR itself. Please refer to the changes in the `docs/` directory.

## Training

N/A

## Certification

N/A

## Marketing

N/A

## Automation tests
 - Unit tests 
   81% code coverage in overall implementation classes
 - Integration tests
   N/A

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Samples

N/A

## Related PRs

The email reminders feature depends on the changes in the following PRs.
- [Required audit data publishers](https://github.com/wso2-extensions/identity-data-publisher-audit/pull/6) 
- [Required artifacts for receiving and persisting audit events in IS Analytics](https://github.com/wso2/analytics-is/pull/399)

## Migrations (if applicable)

N/A

## Test environment

- JDK - 1.8
- Operating Systems - Ubuntu
- Identity Server - 5.5.0
- IS Analytics Server - 5.5.0
 
## Learning

N/A

### CODE REVIEW

File: component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordChangeHandler.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordChangeHandler.java]: Change the year in license header

File: component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordPolicyConstants.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
+/*
+ *  Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordPolicyConstants.java]: Change the year in license header

File: docs/config.md [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+#### Testing the Sample App
+
+Follow the steps given below to test the Password Reset Enforcer.
+
+> These steps should not be followed in a production environment. Therefore if you follow these steps make sure you change it back to the original state accordingly before moving to a production environment.
+
+1. In the `Identity` section under the `Main` tab, click `List` under `Claims`.
+2. Click `http://wso2.org/claims`.
+3. Click `Edit` next to `Last Password Update Time` claim.
```
Comment (Kanapriya) [inline docs/config.md]: Cliam name should be "Last Password Update", not  "Last Password Update Time"

File: component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordResetEnforcer.java [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+    @Override
+    public String getName() {
+        return PasswordPolicyConstants.AUTHENTICATOR_NAME;
+    }
+
+    @Override
+    public AuthenticatorFlowStatus process(HttpServletRequest request, HttpServletResponse response,
+                                           AuthenticationContext context)
+            throws AuthenticationFailedException {
+        // if the logout request comes, then no need to go through and doing complete the flow.
```
Comment (Biruntha) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordResetEnforcer.java]: Start with capital letter

File: component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordResetEnforcer.java [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+                return initiateAuthRequest(response, context, e.getMessage());
+            }
+            return AuthenticatorFlowStatus.SUCCESS_COMPLETED;
+        } else {
+            return initiateAuthRequest(response, context, null);
+        }
+    }
+
+    /**
+     * this will prompt user to change the credentials only if the last password
```
Comment (Biruntha) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordResetEnforcer.java]: Start with capital letter. "This"

File: component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordResetEnforcer.java [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+            String tenantDomain = authenticatedUser.getTenantDomain();
+            String username = authenticatedUser.getAuthenticatedSubjectIdentifier();
+            String tenantAwareUsername = MultitenantUtils.getTenantAwareUsername(username);
+
+            if (hadPasswordExpired(tenantDomain, tenantAwareUsername)) {
+                // The password has expired or the password changed time is not set
+                try {
+                    // Creating the URL to which the user will be redirected
+                    String loginPage = ConfigurationFacade.getInstance().getAuthenticationEndpointURL()
+                            .replace("login.do", "pwd-reset.jsp");
```
Comment (Biruntha) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordResetEnforcer.java]: define a constant and use it for "login.do", "pwd-reset.jsp"

File: component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordChangeHandler.java (hunk 1 above)
Comment (nadundesilva) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordChangeHandler.java]: Changed. (This was a renamed file)

File: component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordPolicyConstants.java (hunk 2 above)
Comment (nadundesilva) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordPolicyConstants.java]: Changed. (This was a renamed file)

File: docs/config.md (hunk 3 above)
Comment (nadundesilva) [reply docs/config.md]: Changed.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordResetEnforcer.java (hunk 4 above)
Comment (nadundesilva) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordResetEnforcer.java]: Changed.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordResetEnforcer.java (hunk 5 above)
Comment (nadundesilva) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordResetEnforcer.java]: Changed.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordResetEnforcer.java (hunk 6 above)
Comment (nadundesilva) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordResetEnforcer.java]: Added.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordResetEnforcer.java [hunk 7]
```diff
@@ -0,0 +1,363 @@
+/*
+ * Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordResetEnforcer.java]: Change the year in license header

File: component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordPolicyUtils.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordPolicyUtils.java]: Change the year in license header

File: component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/internal/PasswordPolicyServiceComponent.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
+/*
+ *  Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/internal/PasswordPolicyServiceComponent.java]: Change the year in license header

File: component/authenticator/src/test/java/org/wso2/carbon/extension/identity/authenticator/passwordpolicy/test/PasswordChangeHandlerTest.java [hunk 10]
```diff
@@ -0,0 +1,390 @@
+/*
+ *  Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
```
Comment (Kanapriya) [inline component/authenticator/src/test/java/org/wso2/carbon/extension/identity/authenticator/passwordpolicy/test/PasswordChangeHandlerTest.java]: Change the year in license header

File: component/authenticator/src/test/java/org/wso2/carbon/extension/identity/authenticator/passwordpolicy/test/PasswordResetEnforcerTest.java [hunk 11]
```diff
@@ -0,0 +1,552 @@
+/*
+ *  Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
```
Comment (Kanapriya) [inline component/authenticator/src/test/java/org/wso2/carbon/extension/identity/authenticator/passwordpolicy/test/PasswordResetEnforcerTest.java]: Change the license header

File: docs/config.md [hunk 12]
```diff
@@ -0,0 +1,200 @@
+# Configuring Password Policy
+
+* [Setting up Password Policy](#setting-up-password-policy)
```
Comment (Kanapriya) [inline docs/config.md]: This link is not working. Can you please fix this as well

File: component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordResetEnforcer.java (hunk 7 above)
Comment (nadundesilva) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordResetEnforcer.java]: Changed.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordPolicyUtils.java (hunk 8 above)
Comment (nadundesilva) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordPolicyUtils.java]: Changed.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/internal/PasswordPolicyServiceComponent.java (hunk 9 above)
Comment (nadundesilva) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/internal/PasswordPolicyServiceComponent.java]: Changed.

File: component/authenticator/src/test/java/org/wso2/carbon/extension/identity/authenticator/passwordpolicy/test/PasswordChangeHandlerTest.java (hunk 10 above)
Comment (nadundesilva) [reply component/authenticator/src/test/java/org/wso2/carbon/extension/identity/authenticator/passwordpolicy/test/PasswordChangeHandlerTest.java]: Changed.

File: component/authenticator/src/test/java/org/wso2/carbon/extension/identity/authenticator/passwordpolicy/test/PasswordResetEnforcerTest.java (hunk 11 above)
Comment (nadundesilva) [reply component/authenticator/src/test/java/org/wso2/carbon/extension/identity/authenticator/passwordpolicy/test/PasswordResetEnforcerTest.java]: Changed.

File: docs/config.md (hunk 12 above)
Comment (nadundesilva) [reply docs/config.md]: Fixed.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordChangeHandler.java [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+        claimMap.put(PasswordPolicyConstants.LAST_CREDENTIAL_UPDATE_TIMESTAMP_CLAIM, Long.toString(timestamp));
+        try {
+            userStoreManager.setUserClaimValues(username, claimMap, null);
+            if (log.isDebugEnabled()) {
+                log.debug("The claim uri "
+                        + PasswordPolicyConstants.LAST_CREDENTIAL_UPDATE_TIMESTAMP_CLAIM + " of "
+                        + username + " updated with the current timestamp");
+            }
+        } catch (UserStoreException e) {
+            log.error("Failed to update claim value for "
```
Comment (hasuniea) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordChangeHandler.java]: If you catch it and log it and rethrow it, then there's no way for the upstream code to know that you've already logged the exception, and so the same exception might get logged twice.

Comment (nadundesilva) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordChangeHandler.java]: Fixed.

**Pull request merged**